nexus/db-queries/src/db/datastore/clickhouse_policy.rs

```diff
@@ -0,0 +1,267 @@
+use diesel::ExpressionMethods;
+use diesel::OptionalExtension;
+use diesel::QueryDsl;
+use nexus_db_model::ClickhouseModeEnum;
+use nexus_db_model::ClickhousePolicy as DbClickhousePolicy;
+use nexus_db_model::DbClickhouseMode;
+use nexus_db_model::SqlU32;
+use nexus_db_model::SqlU8;
+use nexus_types::deployment::ClickhousePolicy;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+
+impl DataStore {
+    pub async fn clickhouse_policy_list(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, SqlU32>,
+    ) -> ListResultVec<ClickhousePolicy> {
+        use db::schema::clickhouse_policy;
+
+        opctx
+            .authorize(authz::Action::ListChildren, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let policies = paginated(
+            clickhouse_policy::table,
+            clickhouse_policy::version,
+            pagparams,
+        )
+        .select(DbClickhousePolicy::as_select())
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(policies.into_iter().map(ClickhousePolicy::from).collect())
+    }
+
+    /// Return the clickhouse policy with the highest version
+    pub async fn clickhouse_policy_get_latest(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Option<ClickhousePolicy>, Error> {
+        opctx.authorize(authz::Action::Read, &authz::BLUEPRINT_CONFIG).await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        use db::schema::clickhouse_policy::dsl;
+
+        let latest_policy = dsl::clickhouse_policy
+            .order_by(dsl::version.desc())
+            .first_async::<DbClickhousePolicy>(&*conn)
+            .await
+            .optional()
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(latest_policy.map(Into::into))
+    }
+
+    /// Insert the current version of the policy in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently
+    /// in the `clickhouse_policy` table. If there are no versions currently
+    /// in the table, then the current policy must be at version 1.
+    pub async fn clickhouse_policy_insert_latest_version(
+        &self,
+        opctx: &OpContext,
+        policy: &ClickhousePolicy,
+    ) -> Result<(), Error> {
+        if policy.version < 1 {
+            return Err(Error::invalid_request(
+                "policy version must be greater than 0",
+            ));
+        }
+        opctx
+            .authorize(authz::Action::Modify, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let num_inserted = if policy.version == 1 {
+            self.clickhouse_policy_insert_first_policy(opctx, &policy).await?
+        } else {
+            let prev_version = policy.version - 1;
```

Comment:
Tiny style nit - since `clickhouse_policy_insert_first_policy` is its own private function, maybe put this query in its own private function too? (Or inline the query for `first_policy`. But it seems weird they're different.)

Reply:
Agreed. Done in [108ec11](https://github.com/oxidecomputer/omicron/pull/6862/commits/108ec119ea0dee25281e6e1f4031d90ed8afb67a)